[CodeGen][NPM] Support CodeGenSCCOrder in pipeline

optimisan opened the pull request:
Wrap passes into Post order CGSCC pass manager in codegen pass builder.

I am adding the pipeline test in this but it is not yet complete.

optimisan committed:
[CodeGen][NPM] Support CodeGenSCCOrder in pipeline
pb/codegenscc-order

optimisan commented:
* **#136818** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136818?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/136818?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

arsenm commented on llvm/test/CodeGen/AMDGPU/llc-pipeline-npm.ll:
```diff
@@ -0,0 +1,144 @@
+; UNSUPPORTED: expensive_checks
```
Why? Is this just from the verifier being on by default? You can use -verify-machineinstrs=0 (and doesn't the -disable-verify just skip the IR verifier on the input)?

arsenm commented on llvm/test/CodeGen/AMDGPU/llc-pipeline-npm.ll:
```diff
@@ -0,0 +1,144 @@
+; UNSUPPORTED: expensive_checks
+; RUN: llc -O3 -enable-new-pm -mtriple=amdgcn--amdhsa -disable-verify -print-pipeline-passes < %s 2>&1 \
+; RUN:   | tr ',' '\n' | FileCheck -check-prefix=GCN-O3 %s
```
Why do you need tr? Does this work on windows

optimisan replied:
Multiple tests use `sed`, so I figured I'll see if `tr` works in the PR check below on windows. It is passing but I don't know how yet, so I can remove `tr` and do something else.

optimisan committed:
Remove tr and support expensive check

arsenm commented on llvm/test/CodeGen/AMDGPU/llc-pipeline-npm.ll:
```diff
@@ -0,0 +1,143 @@
+; RUN: llc -O3 -enable-new-pm -mtriple=amdgcn--amdhsa -disable-verify -print-pipeline-passes < %s 2>&1 \
```
```suggestion
; RUN: llc -enable-new-pm -mtriple=amdgcn--amdhsa -verify-machineinstrs=0 -print-pipeline-passes < %s 2>&1 \
```

optimisan replied:
-verify-machineinstrs doesn't work in NPM, it is replaced by -verify-each which just verifies in the instrumentation.

arsenm commented on llvm/test/CodeGen/AMDGPU/llc-pipeline-npm.ll:
```diff
@@ -0,0 +1,143 @@
+; RUN: llc -O3 -enable-new-pm -mtriple=amdgcn--amdhsa -disable-verify -print-pipeline-passes < %s 2>&1 \
+; RUN:   | FileCheck -check-prefix=GCN-O3 %s
+
+; REQUIRES: asserts
```
Don't think this requires asserts

optimisan committed:
separate tests for O2 and O3

cdevadas commented on llvm/test/CodeGen/AMDGPU/llc-pipeline-npm.ll:
```diff
@@ -1,7 +1,142 @@
-; RUN: llc -O3 -enable-new-pm -mtriple=amdgcn--amdhsa -disable-verify -print-pipeline-passes < %s 2>&1 \
+; RUN: llc -enable-new-pm -disable-verify -mtriple=amdgcn--amdhsa -print-pipeline-passes < %s 2>&1 \
```
Add -O0 as well.

arsenm commented on llvm/test/CodeGen/AMDGPU/llc-pipeline-npm.ll:
```diff
@@ -0,0 +1,278 @@
+; RUN: llc -enable-new-pm -disable-verify -mtriple=amdgcn--amdhsa -print-pipeline-passes < %s 2>&1 \
+; RUN:   | FileCheck -check-prefix=GCN-O2 %s
+
+; RUN: llc -O3 -enable-new-pm -disable-verify -mtriple=amdgcn--amdhsa -print-pipeline-passes < %s 2>&1 \
```
-disable-verify should be removed, it only disables the initial run of the IR verifier on the input

optimisan committed:
condense test into one line

pipeline is printed on a single line, so having CHECK lines on separate
lines can allow extra characters in between (and will not error out on
extra passes being in the pipeline)

arsenm commented on llvm/include/llvm/Passes/CodeGenPassBuilder.h:
```diff
@@ -276,20 +302,47 @@ template <typename DerivedT, typename TargetMachineT> class CodeGenPassBuilder {
+
   private:
+    void flushMFPMToMPM() {
+      if (!MFPM.isEmpty()) {
```
Early return and reduce indent

optimisan committed:
early return

optimisan committed:
more early returns

optimisan committed:
update test

cdevadas reviewed (approved):
LGTM.
